packaging support improvements

rkitover opened the pull request:
This grew out of my work creating an sfml port for macports, but should
be helpful for package maintainers of various distributions:

* add an SFML_USE_SYSTEM_DEPS option to ignore everything in extlibs/
  except for headers/stb_image, and use the system versions

* allow pkg-config file installation if PREFIX/lib/pkgconfig exists or
  a pkg-config program is found, or the SFML_INSTALL_PKGCONFIG_FILES
  flag is set explicitly (this will allow installing pkg-config files
  on e.g. OS X with macports or windows with msys2 etc.)

* on OS X, honor the cmake variables CMAKE_INSTALL_RPATH and
  CMAKE_SKIP_RPATH and only set the INSTALL_NAME_DIR to "@rpath" if
  neither is set

eXpl0it3r commented:
SFMLBuildMaster: Build this please.

mantognini commented:
> [...] SFML_USE_SYSTEM_DEPS [...]

I'm not familiar with macports (brew guy, sorry =P) but what made you prefer this solution (i.e. adding a flag in the CMake script) over specifying paths manually when configuring the project or [removing unwanted directory](https://github.com/Homebrew/homebrew-core/blob/master/Formula/sfml.rb#L35)?

Additionally, here are some of my concern about this option (which was already suggested in the past): https://github.com/SFML/SFML/pull/424#issuecomment-21451792. Maybe you could try to address them?

> [...] SFML_OS_SUPPORTS_PKGCONFIG [...]

This seems like a nice addition for people not using pkgconfig. 👍 

> [...] rpath [...]

I understand the rational for the change but haven't tested it. So, does it still work as one might expect for embedded frameworks/dylibs? I.e. can the user still ship SFML and its dependencies in an app bundle when SFML if build with the default cmake option?

rkitover commented:
@mantognini 

> [...] SFML_USE_SYSTEM_DEPS [...]
I'm not familiar with macports (brew guy, sorry =P) but what made you prefer this solution (i.e. adding a flag in the CMake script) over specifying paths manually when configuring the project or removing unwanted directory?
>
>Additionally, here are some of my concern about this option (which was already suggested in the past): #424 (comment). Maybe you could try to address them?

The approach I'm taking here is different, if the option is set everything in extlibs/ except for stb_image gets added to `CMAKE_IGNORE_PATH` in one place, so no other cmake code needs to be changed or maintained for this, any calls to e.g. `include_directories()` or additions to other cmake paths like `CMAKE_LIBRARY_PATH` will simply be ignored.

>[...] SFML_OS_SUPPORTS_PKGCONFIG [...]
This seems like a nice addition for people not using pkgconfig. 👍

By default the `.pc` files are not installed as it is anyway, so you have to set `SFML_INSTALL_PKGCONFIG_FILES`, the problem is that it only allows you to set the option on Linux or FreeBSD, but there are a lot of other package systems on other OSes (such as macports) that support `pkg-config`.

I'm not sure I did this entirely correctly anyway, it would probably be enough to allow the block to run if the user explicitly set `-DSFML_INSTALL_PKGCONFIG_FILES:BOOL=TRUE` and the `SFML_OS_SUPPORTS_PKGCONFIG` var is probably unnecessary.

The other option would be to default to installing the `.pc` files if `pkg-config` is detected and matches the `CMAKE_INSTALL_PREFIX` instead of requiring that the variable be set, entirely up to you.

>[...] rpath [...]
I understand the rational for the change but haven't tested it. So, does it still work as one might expect for embedded frameworks/dylibs? I.e. can the user still ship SFML and its dependencies in an app bundle when SFML if build with the default cmake option?

Yes by default `@rpath` is set exactly the same way as before, the only difference is if the user (or packaging system) supplied e.g. `-DCMAKE_INSTALL_RPATH=/opt/local/lib` then it would use this instead. `brew` does not have to do this, because `/usr/local/lib` is in the default `@rpath` on OS X, but for macports I had to delete this block in the patches for the current port.

rkitover commented:
@mantognini 

Another thing, currently it refuses to install `pkg-config` files if the libs are built statically, but there is no good reason to not support this, because people will often use calls to `pkg-config` such as:

```bash
pkg-config --static --libs sdl2
```

in the above case you know that the libs are static and you are asking `pkg-config` for all the **other** libs you need to link in order to link the static lib successfully etc..

So I could fix this if you want.

mantognini commented:
> The approach I'm taking here is different, if the option is set everything in extlibs/ except for stb_image gets added to CMAKE_IGNORE_PATH in one place, so no other cmake code needs to be changed or maintained for this, any calls to e.g. include_directories() or additions to other cmake paths like CMAKE_LIBRARY_PATH will simply be ignored.

From this point of view, this actually sounds not that bad. The only issue I've in mind is about the Xcode templates (the post-build script will not copy the right dependencies), but because:
 1. it is already possible to mislead this script (e.g. by using -DFLAC_LIBRARY, or whatever the variable name is), or
 2. no one seems to complain about not being able to use those with brew and macports,

I think it's safe to forget about this for now. If needed it can be addressed in the future with a different PR.

> The other option would be to default to installing the .pc files if pkg-config is detected and matches the CMAKE_INSTALL_PREFIX instead of requiring that the variable be set, entirely up to you.

Full disclosure: I don't use this configuration system nor maintain it here for SFML. So it would be better to let someone with actual experience with pkg-config to decide what's best.

> Yes by default @rpath is set exactly the same way as before

Then this is okay for me.

> Another thing, currently it refuses to install pkg-config files if the libs are built statically, but there is no good reason to not support this
> [...]
> So I could fix this if you want.

Well, if you know how to do it, sure, go ahead. ;-)

rkitover commented:
Just a note, I haven't forgotten about this just some other stuff I need to finish up and I'll try to wrap this up over the next couple of days.

rkitover commented:
@mantognini I've rewritten my commit, and I think it's ready for merging unless you have any other suggestions.

WRT the XCode templates, they are not affected by this code at all (they aren't under extlibs/ .)

eXpl0it3r commented:
SFMLBuildMaster: Build this please.

mantognini commented on tools/pkg-config/sfml-window.pc.in:
```diff
@@ -9,4 +9,6 @@ URL: http://www.sfml-dev.org
 Version: @VERSION_MAJOR@.@VERSION_MINOR@.@VERSION_PATCH@
 Requires: sfml-system
 Libs: -L${libdir} -lsfml-window
+# gl and jpeg may not be in pkg-config
```
Nitpicking: jpeg is not of interest for this file if I'm correct.

mantognini commented on cmake/Macros.cmake:
```diff
@@ -105,9 +105,17 @@ macro(sfml_add_library target)
-                              BUILD_WITH_INSTALL_RPATH 1
-                              INSTALL_NAME_DIR "@rpath")
+        # but only if cmake rpath options aren't set
+        if(NOT CMAKE_SKIP_RPATH AND NOT CMAKE_SKIP_INSTALL_RPATH AND NOT CMAKE_INSTALL_RPATH AND NOT CMAKE_INSTALL_RPATH_USE_LINK_PATH AND NOT CMAKE_INSTALL_NAME_DIR)
```
I'll blindly believe you this is correct!

(I can't believe there are so many variable related to rpath... that's just insane. :-/)

mantognini commented on CMakeLists.txt:
```diff
@@ -130,6 +133,35 @@ if(NOT BUILD_SHARED_LIBS)
+        endwhile()
+    endforeach()
+
+    list(REMOVE_DUPLICATES CMAKE_IGNORE_PATH)
```
Just out of curiosity: without this line, does something break?

rkitover commented:
SFMLBuildMaster: Build this please.

eXpl0it3r commented:
SFMLBuildMaster: Build this please.

rkitover commented:
@mantognini 

I added a check for SFML_INSTALL_XCODE_TEMPLATES if SFML_USE_SYSTEM_DEPS is on, to make them mutually exclusive.

Fixed the comment in the .pc file.

Regarding the REMOVE_DUPLICATES crap, this is because GLOB_RECURSE in cmake can only scan for files, and if you want to directories you have to do this horrible hacky thing.

I used this message from the cmake list for inspiration:

http://public.kitware.com/pipermail/cmake/2012-June/050674.html

mantognini commented:
Awesome, thanks!

All green for merge.

mantognini commented:
@rkitover Any reason why you closed the PR and deleted its content?

rkitover commented:
@mantognini I'm sorry, I was just updating my local sources to update the macports package etc. and thought it was merged already, once I pushed upstream on top of my fork github automatically closed the PR, I restored the commit but it won't let me reopen it.

mantognini commented:
All good then. :-)

eXpl0it3r commented:
SFMLBuildMaster: Build this please.

eXpl0it3r commented:
Merge in 5fe5e5d6d7792e37685a437551ffa8ed5161fcc1 on master.
